plugin/vscode/performance-writer.js

```diff
@@ -0,0 +1,113 @@
+// Copyright (C) 2020  Matthew Glazar
+// See end of file for extended copyright information.
+
+const fs = require("fs");
+const path = require("path");
+const process = require("process");
+
+class PerformanceWriter {
+  constructor(extensionContext) {
+    this._extensionContext = extensionContext;
+    this._dataAttributes = {
+      columns: ["timestamp", "file-ref", "event-id", "file-name-or-duration"],
+    };
+    this._tmpLogLines = [];
+    this._fileIndex = 0;
+    this._writeToFileTimer = null;
+    this._timerActivated = false;
+    if (!fs.existsSync(this._extensionContext.logUri.fsPath)) {
+      fs.mkdirSync(this._extensionContext.logUri.fsPath);
+      this.writeHeaderLine();
+    }
+    this._performanceTxtPath = path.join(
+      this._extensionContext.logUri.fsPath,
+      "qljsPerformanceTrace.txt"
+    );
+    this._writableStream = fs.createWriteStream(this._performanceTxtPath, {
+      encoding: "utf8",
+      flags: "a",
+    });
+    this._currentSessionMap = {};
+  }
+
+  fileNameOptimizer(fileName) {
+    if (
+      Object.prototype.hasOwnProperty.call(this._currentSessionMap, fileName)
+    ) {
+      this._fileIndex = this._currentSessionMap[fileName];
+    } else {
+      this._fileIndex++;
+      this._currentSessionMap[fileName] = this._fileIndex;
+      this.writeNewFileLine(fileName);
+    }
+  }
+
+  initiateLogging(timeTaken, fileName) {
+    this.fileNameOptimizer(fileName);
+    this.writeExistingFileLine(timeTaken);
+    this.setWriteToFileTimer();
+  }
+
+  writeHeaderLine() {
+    this._tmpLogLines.push(...this._dataAttributes.columns);
+  }
+
+  writeNewFileLine(fileName) {
+    this._tmpLogLines.push(Date.now(), this._fileIndex, "n", fileName);
+  }
+
+  writeExistingFileLine(timeTaken) {
+    this._tmpLogLines.push(Date.now(), this._fileIndex, "p", timeTaken);
+  }
+
+  setWriteToFileTimer() {
+    if (!this._timerActivated) {
+      this._timerActivated = true;
+      this._writeToFileTimer = setInterval(() => {
```

Comment:
Why setInterval? I think the logic would be simpler with setTimeout. Maybe not, though.

Reply:
Current logic is as follows for logging linting performance:
setWriteToFileTimer is activated every single time a key stroke is pressed => setInterval is activated => keeps going until the tmpLogLines is empty AND also it is locked from being activated again via the "!this._timerActivated" condition. 

If using setTimeout , the logic, in my opinion would be exactly the same because you'll need to activate setTimeout again if there is new tmpLogLines while waiting for the setTimeout to activate. I'll need to build out some sort of logic to know to setTimeout again if there is new tmpLogLines elements while waiting for setTimeout to activate which is essentially setInterval.

On the other hand, you could just activate setTimeout everytime a keystroke is pressed and you'll have _many_ setTimeout functions going on... but I'm sure that's not what you meant with this implementation. 

Conclusion: we should probably just keep setInterval as setTimeout would boil down to the same logic imo.

Reply:
> If using setTimeout , the logic, in my opinion would be exactly the same because you'll need to activate setTimeout again if there is new tmpLogLines while waiting for the setTimeout to activate.

No, you wouldn't need to activate setTimeout again if there is new tmpLogLines. You can write those tmpLogLines immediately.

Reply:
Fixed in latest commit:
Did it as you suggested